Add disable support to PostStatus settings panel

noahtallen opened the pull request:
While many other panels support the isEditorPanelEnabled option, the `PostStatus` panel does not.  In my own project, I was trying to use `dispatch('core/edit-post').removeEditorPanel( 'post-status' );` to remove this panel in a particular context. I realized that it wasn't working and found that the component did not respect the setting.

Note: I have a PR up here which is the context of finding this bug: https://github.com/Automattic/wp-calypso/pull/35643

## Description
-  If the 'post-status' panel is not enabled, the component now returns null. Previously, it always returned the full Panel regardless of whether it was enabled. I used the prior art of the featured-image panel to make these changes.

## How has this been tested?
1. I tested this fix with my own project and `removeEditorPanel( 'post-status' );` now works correctly from a JS context. (This was using Gutenberg master on the Gutenberg docker setup.)
2. I tried `wp.data.dispatch('core/edit-post').removeEditorPanel( 'post-status' );` from the console on a normal post (outside of my project), and it works correctly. I would recommend this option in order to test this branch.

## Types of changes
- Bug fix.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

noahtallen committed:
Add disable support to PostStatus settings panel

While many other panels support the isEditorPanelEnabled option, the
PostStatus panel does not. If the 'post-status' panel is not enabled, the
component now returns null. Previously, it always returned the full
Panel regardless of whether it was enabled.

This is cleaner than returning null, since ifCondition does that for us.

swissspidy reviewed (commented):
Could use the ifCondition HOC here.

youknowriad commented:
Do we know why this has been excluded initially? Other than that, this LGTM

noahtallen commented:
I pinged the guy who originally added support and he directed me to this comment: https://github.com/WordPress/gutenberg/pull/10215#issuecomment-429795018

In particular, they didn't want to allow a user to remove the panel from a checkbox in the Options modal. This PR doesn't add a checkbox for it, so I think it should still be ok since we're only adding developer support.

> ifCondition

I'll check that out! It sounds nifty. I think I saw it in use on one of the other panel components. 🤔

youknowriad commented:
Ok adding @mtias to the discussion because it was his decision.

noahtallen commented:
Some additional context for our situation is that we have a separate editor for header/footer. In that editor, the user can see the PostStatus panel even though the visibility/publish/trash shouldn't be changed by the end user.

gwwar commented on packages/edit-post/src/components/sidebar/post-status/index.js:
```diff
@@ -4,7 +4,7 @@
 import { PanelBody } from '@wordpress/components';
 import { withSelect, withDispatch } from '@wordpress/data';
-import { compose } from '@wordpress/compose';
+import { compose, ifCondition } from '@wordpress/compose';
```
Fancy. TIL ✨ 

https://github.com/WordPress/gutenberg/blob/da0b1a358a1cf58834c9f48d55b878e0a69167dc/packages/compose/src/higher-order/if-condition/README.md

noahtallen replied:
so many neat packages in this repo!

youknowriad commented on packages/edit-post/src/components/sidebar/post-status/index.js:
```diff
@@ -45,13 +45,18 @@ function PostStatus( { isOpened, onTogglePanel } ) {
 			return dispatch( 'core/edit-post' ).toggleEditorPanelOpened( PANEL_NAME );
 		},
 	} ) ),
+	ifCondition( ( { isEnabled } ) => isEnabled ),
```
It could even be moved before withDispatch. Probably we won't gain anything perceivable but you know bail as early as possible :P

noahtallen replied:
Done ;)

gwwar commented:
This tests well for me, and code makes sense. Sounds like we probably just need a confirmation on if we want to allow this or not in the project.

![post-status](https://user-images.githubusercontent.com/1270189/63462474-e1998900-c40f-11e9-9b61-f77f22d19732.gif)

gwwar commented:
@noahtallen maybe rebase to see if that clears up the tests? Might have picked up a bad snapshot of the npm version in travis.

noahtallen commented:
Need to work on my Github forking skills. Will do :)

noahtallen commented:
Looks like the build was passing for a sec while I was rebasing, but is failing again. Everything seems to be testing ok locally, so I'm wondering if this is a flakey test 🤔I can't rerun the build myself, so I thought I'd check if this is common :)

youknowriad commented:
Looks like there might be an intermitent faiilure in the list block test (keyboard to indent/outdent). Pinging @gziolo @ellatrix in case they know more about it.

I restarted the job to see.

noahtallen commented:
> I restarted the job to see.

Looks like it passed!

gziolo reviewed (changes requested):
To programmatically remove editor panels from UI you should use `removeEditorPanel ` action together with `isEditorPanelRemoved` selector implemented in #11802.

`isEditorPanelEnabled` isn't a good fit here as it combines also the state from UI. In theory, it works as it checks both conditions behind the scenes but as explained by @mtias we don't want to allow to disable the post status panel through UI.

Once, this is updated, this PR is good to go :)

gziolo commented:
It would be great to find a good place in docs where we document how developers can programmatically remove all different panels. I opened #17281 to track it.

noahtallen committed:
Use isEditorPanelRemoved instead of isEditorPanelEnabled

We do this since the panel should never be disabled with the enabled
selector, but it can be programmatically removed.

noahtallen commented:
> use removeEditorPanel action together with isEditorPanelRemoved selector

Sounds great! Updated to reflect that.

gziolo commented on packages/edit-post/src/components/sidebar/post-status/index.js:
```diff
@@ -45,9 +45,16 @@ function PostStatus( { isOpened, onTogglePanel } ) {
+			isOpened: isEditorPanelOpened( PANEL_NAME ),
+		};
+	} ),
+	ifCondition( ( { isRemoved } ) => isRemoved ),
```
This logic should be updated:

```suggestion
	ifCondition( ( { isRemoved } ) => ! isRemoved ),
```

Travis reports some failing tests and thy are valid.

gziolo replied:
I can't push to your repository myself :(

I tried with the proposed change locally and it works properly.

noahtallen replied:
> I can't push to your repository myself

Likewise ;)

Great catch! Added that.

gziolo commented:
Should we update the following e2e test to validate that `Status & Visibility` panel can be removed as well? See:
https://github.com/WordPress/gutenberg/blob/b786aad41b4670190b6cac4d99bab3abd3f85049/packages/e2e-tests/specs/sidebar.test.js#L113-L133

noahtallen committed:
Update logic for removing post status

noahtallen committed:
Update e2e test to remove PostStatus panel

noahtallen commented:
> Should we update the following e2e test

Awesome! Just updated that and added a check for 'Status & Visibility'.

gziolo commented on packages/e2e-tests/specs/sidebar.test.js:
```diff
@@ -124,12 +124,14 @@ describe( 'Sidebar', () => {
 			removeEditorPanel( 'featured-image' );
```
It would be great to add a check a few lines earlier whether `Status & Visibility` is there before it gets removed:

```js
expect( await findSidebarPanelWithTitle( 'Status & Visibility' ) ).toBeDefined();
```

gziolo reviewed (approved):
Nice work, thanks 👍 

There is still one line missing in the test as noted in the comment

noahtallen committed:
Update e2e test to check if Status Panel exists

noahtallen commented:
Updated the test to handle that condition. :)

gziolo commented:
@noahtallen - thanks for addressing all feedback and congrats on your first code contribution to Gutenberg 🎉
